Kernel CVE Reporting bundle

jakub-pomykala opened the pull request.

jakub-pomykala committed:
Creating ServerConfiguration for FixPackReporting

- Changes to CVEServiceClient and CVEResponseHandler for managing JSONArray response from tests
- Renaming FATs
- Adding in further tests
- Changing behaviour when incorrect URL is provided
- Created new custom exception (DataCollectorException)
- Created buildExceptionMessage()
- Created unit tests for buildExceptionMessage() and buildJsonString()
- Split and renamed CVEServiceClient.fixPackData() to retrieveCVEData()

Making tests use a keystore to allow https connection

Initial commit of CVE Reporting & Insight feature

- Initial code for CVE Reporting & Insight
- Removing com.ibm.json4j from kernel-1.0.mf and corresponding features as moving it to kernelCore-1.0.mf for use in CVE Reporting & insight feature
- Initial FAT tests for CVE Reporting & Insights

Resolving issue with feature running (fix in bnd)

- Updating names in ServerConfigurations.java

Applying code suggestion from Jared

- Mostly copyright tag fixes

Co-authored-by: Jared Anderson <jhanders@us.ibm.com>

Moving ifEnabled() from CheckIfEnabledUtils

- Moved ifEnabled() into FixReportingComponent class (as only used there)
- Updated the messages in nlsprops
- Added handling for the url link that will be returned from the cloud service
- Updated javadoc for retrieveCVEData()
- Moved and changed the condition verifying if url link starts with https protocol
- Updated getResponse() and sendData() to use try-with-resource
- Updated the buildJsonString() to use just a StringBuilder()
- Removing the writeObject() from dataCollectorException class as this will never have NotSerializableException()
- Updating the tearDown() in CVEResponseTest class to use finally for the secondary server
-

Adding comma to more.information.message

Update CVEReporting.nlsprops

jhanders34 commented on dev/com.ibm.ws.kernel.reporting/.project:
```diff
@@ -0,0 +1,34 @@
+				<arguments>node_modules|\.git|__CREATED_BY_JAVA_LANGUAGE_SERVER__</arguments>
+			</matcher>
+		</filter>
+	</filteredResources>
```
```suggestion
```

jhanders34 commented on dev/com.ibm.ws.kernel.reporting/bnd.bnd:
```diff
@@ -0,0 +1,57 @@
+#*******************************************************************************
+# Copyright (c) 2017, 2024 IBM Corporation and others.
```
```suggestion
# Copyright (c) 2024 IBM Corporation and others.
```

jhanders34 commented on dev/com.ibm.ws.kernel.reporting/bnd.bnd:
```diff
@@ -0,0 +1,57 @@
+# SPDX-License-Identifier: EPL-2.0
+#
+# Contributors:
+#     IBM Corporation - initial API and implementation
```
```suggestion
```

Not needed in new files.

jhanders34 commented on dev/com.ibm.ws.kernel.reporting/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -0,0 +1,21 @@
+# http://www.eclipse.org/legal/epl-v10.html
+#
+# Contributors:
+#     IBM Corporation - initial API and implementation
```
```suggestion
# are made available under the terms of the Eclipse Public License 2.0
# which accompanies this distribution, and is available at
# http://www.eclipse.org/legal/epl-2.0/
#
# SPDX-License-Identifier: EPL-2.0
```

jhanders34 commented on dev/com.ibm.ws.kernel.reporting/resources/OSGI-INF/metatype/metatype.xml:
```diff
@@ -0,0 +1,43 @@
+    http://www.eclipse.org/legal/epl-v10.html
+   
+    Contributors:
+        IBM Corporation - initial API and implementation
```
```suggestion
    are made available under the terms of the Eclipse Public License 2.0
    which accompanies this distribution, and is available at
    http://www.eclipse.org/legal/epl-2.0/

    SPDX-License-Identifier: EPL-2.0
```

jhanders34 commented on dev/com.ibm.ws.kernel.reporting/resources/com/ibm/ws/kernel/reporting/internal/resources/CVEReporting.nlsprops:
```diff
@@ -0,0 +1,59 @@
+# SPDX-License-Identifier: EPL-2.0
+# 
+# Contributors:
+#     IBM Corporation - initial API and implementation
```
```suggestion
```

jhanders34 commented on dev/com.ibm.ws.kernel.reporting/src/com/ibm/ws/kernel/reporting/CVEResponseHandler.java:
```diff
@@ -0,0 +1,77 @@
+package com.ibm.ws.kernel.reporting;
```
```suggestion
/*******************************************************************************
 * Copyright (c) 2024 IBM Corporation and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *******************************************************************************/

package com.ibm.ws.kernel.reporting;
```

jhanders34 commented on dev/com.ibm.ws.kernel.reporting/src/com/ibm/ws/kernel/reporting/CVEServiceClient.java:
```diff
@@ -0,0 +1,161 @@
+package com.ibm.ws.kernel.reporting;
```
```suggestion
/*******************************************************************************
 * Copyright (c) 2024 IBM Corporation and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *******************************************************************************/

package com.ibm.ws.kernel.reporting;
```

jhanders34 commented on dev/com.ibm.ws.kernel.reporting/src/com/ibm/ws/kernel/reporting/DataCollector.java:
```diff
@@ -0,0 +1,152 @@
+package com.ibm.ws.kernel.reporting;
```
```suggestion
/*******************************************************************************
 * Copyright (c) 2024 IBM Corporation and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *******************************************************************************/

package com.ibm.ws.kernel.reporting;
```

jhanders34 commented on dev/com.ibm.ws.kernel.reporting/src/com/ibm/ws/kernel/reporting/DataCollectorException.java:
```diff
@@ -0,0 +1,27 @@
+package com.ibm.ws.kernel.reporting;
```
```suggestion
/*******************************************************************************
 * Copyright (c) 2024 IBM Corporation and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *******************************************************************************/

package com.ibm.ws.kernel.reporting;
```

jhanders34 commented on dev/com.ibm.ws.kernel.reporting/src/com/ibm/ws/kernel/reporting/FixReportingComponent.java:
```diff
@@ -0,0 +1,123 @@
+
+		reporterTask = new ReporterTask(featureProvisioner, fixManager, serverInfo, properties);
+
+		if (System.getProperty("set.reporting.to.working") != null && Boolean.getBoolean("set.reporting.to.working")) {
```
Java security needs to be taken into consideration.  This may not be required if this is early enough in server startup, but I am guessing it might not be.  So you need to do the AccessController.doPriviledged in order get the system property.

jakub-pomykala replied:
Addressed.

jhanders34 commented on dev/com.ibm.ws.kernel.reporting/src/com/ibm/ws/kernel/reporting/ReporterTask.java:
```diff
@@ -0,0 +1,102 @@
+package com.ibm.ws.kernel.reporting;
```
```suggestion
/*******************************************************************************
 * Copyright (c) 2024 IBM Corporation and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *******************************************************************************/

package com.ibm.ws.kernel.reporting;
```

jhanders34 commented on dev/com.ibm.ws.kernel.reporting/src/com/ibm/ws/kernel/reporting/package-info.java:
```diff
@@ -0,0 +1,21 @@
+/*******************************************************************************
+ * Copyright (c) 2009 IBM Corporation and others.
```
```suggestion
 * Copyright (c) 2024 IBM Corporation and others.
```

jhanders34 commented on dev/com.ibm.ws.kernel.reporting/src/com/ibm/ws/kernel/reporting/package-info.java:
```diff
@@ -0,0 +1,21 @@
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
```
```suggestion
```

jhanders34 commented on dev/com.ibm.ws.kernel.reporting/src/com/ibm/ws/kernel/reporting/utils/CheckIfEnabledUtils.java:
```diff
@@ -0,0 +1,27 @@
+package com.ibm.ws.kernel.reporting.utils;
```
```suggestion
/*******************************************************************************
 * Copyright (c) 2024 IBM Corporation and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *******************************************************************************/

package com.ibm.ws.kernel.reporting.utils;
```

jhanders34 commented on dev/com.ibm.ws.kernel.reporting/src/com/ibm/ws/kernel/reporting/utils/CheckIfEnabledUtils.java:
```diff
@@ -0,0 +1,27 @@
+
+import java.util.Map;
+
+public class CheckIfEnabledUtils {
```
This is only used in the one class.  Can the method just be moved to that one class instead of having a separate class for just one method?

jakub-pomykala replied:
Moved ifEnabled() into FixReportingComponent class

jhanders34 commented on dev/com.ibm.ws.kernel.reporting/test/com/ibm/ws/kernel/reporting/CreateJsonStringTest.java:
```diff
@@ -0,0 +1,63 @@
+package com.ibm.ws.kernel.reporting;
```
```suggestion
/*******************************************************************************
 * Copyright (c) 2024 IBM Corporation and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *******************************************************************************/

package com.ibm.ws.kernel.reporting;
```

jhanders34 commented on dev/com.ibm.ws.kernel.reporting/test/com/ibm/ws/kernel/reporting/ReporterTaskTest.java:
```diff
@@ -0,0 +1,29 @@
+package com.ibm.ws.kernel.reporting;
```
```suggestion
/*******************************************************************************
 * Copyright (c) 2024 IBM Corporation and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *******************************************************************************/

package com.ibm.ws.kernel.reporting;
```

jhanders34 commented on dev/com.ibm.ws.kernel.reporting_fat/.classpath:
```diff
@@ -0,0 +1,12 @@
+		<attributes>
+			<attribute name="module" value="true"/>
+		</attributes>
+	</classpathentry>
```
```suggestion
	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-1.8"/>
```

jhanders34 commented on dev/com.ibm.ws.kernel.reporting_fat/bnd.bnd:
```diff
@@ -0,0 +1,39 @@
+#*******************************************************************************
+# Copyright (c) 2020, 2023 IBM Corporation and others.
```
```suggestion
# Copyright (c) 2024 IBM Corporation and others.
```

jhanders34 commented on dev/com.ibm.ws.kernel.reporting_fat/bnd.bnd:
```diff
@@ -0,0 +1,39 @@
+# SPDX-License-Identifier: EPL-2.0
+#
+# Contributors:
+#     IBM Corporation - initial API and implementation
```
```suggestion
```

jhanders34 commented on dev/com.ibm.ws.kernel.reporting_fat/build.gradle:
```diff
@@ -0,0 +1,41 @@
+configurations {
```
```suggestion
/*******************************************************************************
 * Copyright (c) 2024 IBM Corporation and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *******************************************************************************/

configurations {
```

jhanders34 commented on dev/com.ibm.ws.kernel.reporting_fat/fat/src/com/ibm/ws/kernel/reporting/fat/CVEDataTest.java:
```diff
@@ -0,0 +1,130 @@
+package com.ibm.ws.kernel.reporting.fat;
```
```suggestion
/*******************************************************************************
 * Copyright (c) 2024 IBM Corporation and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *******************************************************************************/

package com.ibm.ws.kernel.reporting.fat;
```

jhanders34 commented on dev/com.ibm.ws.kernel.reporting_fat/fat/src/com/ibm/ws/kernel/reporting/fat/CVEReportingConfigTest.java:
```diff
@@ -0,0 +1,106 @@
+package com.ibm.ws.kernel.reporting.fat;
```
```suggestion
/*******************************************************************************
 * Copyright (c) 2024 IBM Corporation and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *******************************************************************************/

package com.ibm.ws.kernel.reporting.fat;
```

jhanders34 commented on dev/com.ibm.ws.kernel.reporting_fat/fat/src/com/ibm/ws/kernel/reporting/fat/CVEResponseTest.java:
```diff
@@ -0,0 +1,184 @@
+package com.ibm.ws.kernel.reporting.fat;
```
```suggestion
/*******************************************************************************
 * Copyright (c) 2024 IBM Corporation and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *******************************************************************************/

package com.ibm.ws.kernel.reporting.fat;
```

jhanders34 commented on dev/com.ibm.ws.kernel.reporting_fat/fat/src/com/ibm/ws/kernel/reporting/fat/FATSuite.java:
```diff
@@ -0,0 +1,11 @@
+package com.ibm.ws.kernel.reporting.fat;
```
```suggestion
/*******************************************************************************
 * Copyright (c) 2024 IBM Corporation and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *******************************************************************************/

package com.ibm.ws.kernel.reporting.fat;
```

jhanders34 commented on dev/com.ibm.ws.kernel.reporting_fat/fat/src/com/ibm/ws/kernel/reporting/fat/response/CVEReportingResponseEndpoints.java:
```diff
@@ -0,0 +1,39 @@
+package com.ibm.ws.kernel.reporting.fat.response;
```
```suggestion
/*******************************************************************************
 * Copyright (c) 2024 IBM Corporation and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *******************************************************************************/

package com.ibm.ws.kernel.reporting.fat.response;
```

jhanders34 commented on dev/com.ibm.ws.kernel.reporting_fat/fat/src/com/ibm/ws/kernel/reporting/fat/response/JsonData.java:
```diff
@@ -0,0 +1,11 @@
+package com.ibm.ws.kernel.reporting.fat.response;
```
```suggestion
/*******************************************************************************
 * Copyright (c) 2024 IBM Corporation and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *******************************************************************************/

package com.ibm.ws.kernel.reporting.fat.response;
```

jhanders34 commented on dev/com.ibm.ws.kernel.reporting_fat/publish/files/test.InterimFixes-1.0.mf:
```diff
@@ -0,0 +1,6 @@
+Subsystem-ManifestVersion: 1.0
+Subsystem-SymbolicName: test.InterimFixes-1.0;visibility:=public
+Subsystem-Version: 1.0.0
+Subsystem-Content: ProvisioningInterimFixesTestBundle1; version="[1,1.0.100)"
```
```suggestion
Subsystem-Content: ProvisioningInterimFixesTestBundle1; version="[1,1.1)"
```

jhanders34 commented on dev/com.ibm.ws.kernel.reporting_fat/publish/files/testFeatureListServer.xml:
```diff
@@ -0,0 +1,24 @@
+<server description="Server for testing CVE reporting">
```
```suggestion
<!--
    Copyright (c) 2024 IBM Corporation and others.
    All rights reserved. This program and the accompanying materials
    are made available under the terms of the Eclipse Public License 2.0
    which accompanies this distribution, and is available at
    http://www.eclipse.org/legal/epl-2.0/

    SPDX-License-Identifier: EPL-2.0
 -->

<server description="Server for testing CVE reporting">
```

jhanders34 commented on dev/com.ibm.ws.kernel.reporting_fat/publish/servers/com.ibm.ws.kernel.reporting.response.server/bootstrap.properties:
```diff
@@ -0,0 +1 @@
+bootstrap.include=../testports.properties
```
```suggestion
###############################################################################
# Copyright (c) 2024 IBM Corporation and others.
# All rights reserved. This program and the accompanying materials
# are made available under the terms of the Eclipse Public License 2.0
# which accompanies this distribution, and is available at
# http://www.eclipse.org/legal/epl-2.0/
#
# SPDX-License-Identifier: EPL-2.0
###############################################################################
bootstrap.include=../testports.properties
```

jhanders34 commented on dev/com.ibm.ws.kernel.reporting_fat/publish/servers/com.ibm.ws.kernel.reporting.response.server/jvm.options:
```diff
@@ -0,0 +1,2 @@
+-Dcom.ibm.ws.beta.edition=true
```
```suggestion
###############################################################################
# Copyright (c) 2024 IBM Corporation and others.
# All rights reserved. This program and the accompanying materials
# are made available under the terms of the Eclipse Public License 2.0
# which accompanies this distribution, and is available at
# http://www.eclipse.org/legal/epl-2.0/
#
# SPDX-License-Identifier: EPL-2.0
###############################################################################
-Dcom.ibm.ws.beta.edition=true
```

jhanders34 commented on dev/com.ibm.ws.kernel.reporting_fat/publish/servers/com.ibm.ws.kernel.reporting.response.server/server.xml:
```diff
@@ -0,0 +1,35 @@
+<server description="Server for testing CVE reporting (app server)">
```
```suggestion
<!--
    Copyright (c) 2024 IBM Corporation and others.
    All rights reserved. This program and the accompanying materials
    are made available under the terms of the Eclipse Public License 2.0
    which accompanies this distribution, and is available at
    http://www.eclipse.org/legal/epl-2.0/

    SPDX-License-Identifier: EPL-2.0
 -->
 
 <server description="Server for testing CVE reporting (app server)">
```

jhanders34 commented on dev/com.ibm.ws.kernel.reporting_fat/publish/servers/com.ibm.ws.kernel.reporting.response.test.server/server.xml:
```diff
@@ -0,0 +1,30 @@
+<server description="Server for testing CVE reporting (kernel server)">
```
```suggestion
<!--
    Copyright (c) 2024 IBM Corporation and others.
    All rights reserved. This program and the accompanying materials
    are made available under the terms of the Eclipse Public License 2.0
    which accompanies this distribution, and is available at
    http://www.eclipse.org/legal/epl-2.0/

    SPDX-License-Identifier: EPL-2.0
 -->
 
 <server description="Server for testing CVE reporting (kernel server)">
```

jhanders34 commented on dev/com.ibm.ws.kernel.reporting_fat/publish/servers/com.ibm.ws.kernel.reporting.response.test.server/bootstrap.properties:
```diff
@@ -0,0 +1 @@
+bootstrap.include=../testports.properties
```
```suggestion
###############################################################################
# Copyright (c) 2024 IBM Corporation and others.
# All rights reserved. This program and the accompanying materials
# are made available under the terms of the Eclipse Public License 2.0
# which accompanies this distribution, and is available at
# http://www.eclipse.org/legal/epl-2.0/
#
# SPDX-License-Identifier: EPL-2.0
###############################################################################
bootstrap.include=../testports.properties
```

jhanders34 commented on dev/com.ibm.ws.kernel.reporting_fat/publish/servers/com.ibm.ws.kernel.reporting.server/bootstrap.properties:
```diff
@@ -0,0 +1 @@
+bootstrap.include=../testports.properties
```
```suggestion
###############################################################################
# Copyright (c) 2024 IBM Corporation and others.
# All rights reserved. This program and the accompanying materials
# are made available under the terms of the Eclipse Public License 2.0
# which accompanies this distribution, and is available at
# http://www.eclipse.org/legal/epl-2.0/
#
# SPDX-License-Identifier: EPL-2.0
###############################################################################
bootstrap.include=../testports.properties
```

jhanders34 commented on dev/com.ibm.ws.kernel.reporting_fat/publish/servers/com.ibm.ws.kernel.reporting.server/jvm.options:
```diff
@@ -0,0 +1,2 @@
+-Dcom.ibm.ws.beta.edition=true
```
```suggestion
###############################################################################
# Copyright (c) 2024 IBM Corporation and others.
# All rights reserved. This program and the accompanying materials
# are made available under the terms of the Eclipse Public License 2.0
# which accompanies this distribution, and is available at
# http://www.eclipse.org/legal/epl-2.0/
#
# SPDX-License-Identifier: EPL-2.0
###############################################################################
-Dcom.ibm.ws.beta.edition=true
```

jhanders34 commented on dev/com.ibm.ws.kernel.reporting_fat/publish/servers/com.ibm.ws.kernel.reporting.server/server.xml:
```diff
@@ -0,0 +1,21 @@
+<server description="Server for testing CVE reporting">
```
```suggestion
<!--
    Copyright (c) 2024 IBM Corporation and others.
    All rights reserved. This program and the accompanying materials
    are made available under the terms of the Eclipse Public License 2.0
    which accompanies this distribution, and is available at
    http://www.eclipse.org/legal/epl-2.0/

    SPDX-License-Identifier: EPL-2.0
 -->

<server description="Server for testing CVE reporting">
```

jhanders34 commented on dev/com.ibm.ws.kernel.reporting/bnd.bnd:
```diff
@@ -0,0 +1,57 @@
+ OSGI-INF/metatype/metatype.xml=resources/OSGI-INF/metatype/metatype.xml
+
+#Import-Package: !com.sun.management, !com.ibm.lang.management, com.ibm.json4j.java;version="[1.0.0,2.0.0)",* 
+Import-Package: !com.sun.management, !com.ibm.lang.management, *
```
```suggestion
```
These shouldn't be needed

jhanders34 commented on dev/com.ibm.ws.kernel.reporting/bnd.bnd:
```diff
@@ -0,0 +1,57 @@
+
+-dsannotations: com.ibm.ws.kernel.reporting.FixReportingComponent
+
+instrument.classesExcludes: io/**
```
What `io/**` classes?  This should be the nls properties classes.

jakub-pomykala replied:
Replaced with `instrument.classesExcludes: com/ibm/ws/kernel/reporting/internal/resources/*.class`

jhanders34 commented on dev/com.ibm.ws.kernel.reporting/src/com/ibm/ws/kernel/reporting/CVEResponseHandler.java:
```diff
@@ -0,0 +1,77 @@
+package com.ibm.ws.kernel.reporting;
```
This is a new project / package.  As such the package should be prefixed with io.openliberty instead of com.ibm.ws.  Same for the nlsprops file

jakub-pomykala replied:
Done.

jhanders34 commented on dev/com.ibm.ws.kernel.reporting/src/com/ibm/ws/kernel/reporting/CVEResponseHandler.java:
```diff
@@ -0,0 +1,77 @@
+				Tr.warning(tc, "CWWKF1702.cve.found", cves.toString());
+			} else {
+				Tr.warning(tc, "CWWKF1702.cve.found", cves.toString());
+			}
```
The if and the else do the same thing.  What is the reason for this conditional?  Is the else supposed to do something different?

jakub-pomykala replied:
Removed the block

jhanders34 commented on dev/com.ibm.ws.kernel.reporting/src/com/ibm/ws/kernel/reporting/CVEServiceClient.java:
```diff
@@ -0,0 +1,161 @@
+	 * @param data    Map<String, String>
+	 * @param urlLink URL link which is set in the Server.xml
+	 * @return responseFromService String
+	 * @throws Exception
```
Javadoc doesn't match the method signature

jakub-pomykala replied:
Fixed

jhanders34 commented on dev/com.ibm.ws.kernel.reporting/src/com/ibm/ws/kernel/reporting/CVEServiceClient.java:
```diff
@@ -0,0 +1,161 @@
+		String link = urlLink;
+		URL url = null;
+		url = new URL(link);
+		if (!url.getProtocol().equals("https")) {
```
This checking should be first.  No reason to create the things before this check if the exception is thrown.  The exception maybe should say expecting https.  Right now it says that the protocol is invalid, but doesn't say what the valid option is.

jakub-pomykala replied:
For testability of the code I left the `String jsonData = buildJsonString(data);` to be done before the check but moved everything else to be done after the check is carried out.

jhanders34 commented on dev/com.ibm.ws.kernel.reporting/src/com/ibm/ws/kernel/reporting/CVEServiceClient.java:
```diff
@@ -0,0 +1,161 @@
+	 */
+	private static void sendData(HttpsURLConnection connection, String jsonData) throws ConnectException, IOException {
+		OutputStream os = connection.getOutputStream();
+		byte[] input = jsonData.getBytes("utf-8");
```
```suggestion
		byte[] input = jsonData.getBytes(StandardCharsets.UTF_8);
```

jhanders34 commented on dev/com.ibm.ws.kernel.reporting/src/com/ibm/ws/kernel/reporting/CVEServiceClient.java:
```diff
@@ -0,0 +1,161 @@
+	 * @throws IOException
+	 */
+	private static StringBuilder getResponse(HttpsURLConnection connection, StringBuilder response) throws IOException {
+		BufferedReader br = new BufferedReader(new InputStreamReader(connection.getInputStream(), "utf-8"));
```
```suggestion
		BufferedReader br = new BufferedReader(new InputStreamReader(connection.getInputStream(), StandardCharsets.UTF_8));
```

jhanders34 replied:
This should also be in a try with resources so that it closes the reader?  Not sure if that will close the connection input stream and if that is what we want or not.

jakub-pomykala replied:
This and the `sendData()` are now in a try-with-resources.

jhanders34 commented on dev/com.ibm.ws.kernel.reporting/src/com/ibm/ws/kernel/reporting/CVEServiceClient.java:
```diff
@@ -0,0 +1,161 @@
+	 * @return A StringBuilder with the response from the cloud service
+	 * @throws IOException
+	 */
+	private static StringBuilder getResponse(HttpsURLConnection connection, StringBuilder response) throws IOException {
```
Not sure why you are taking the StringBuilder in and then returning it as well.  Seems this method could be void or not have a StringBuilder parameter and just return String and have the method make the StringBuilder.

jakub-pomykala replied:
Updated to now return a string.

jhanders34 commented on dev/com.ibm.ws.kernel.reporting/src/com/ibm/ws/kernel/reporting/CVEServiceClient.java:
```diff
@@ -0,0 +1,161 @@
+	 * @return A string
+	 */
+	protected String buildJsonString(Map<String, String> data) {
+		String jsonData;
```
Should be using a StringBuilder here.  When doing String concatenation in this method, it is just making a new StringBuilder each time, so you end up with lots and lots of StringBulder and lots and lots of Strings for no value add.

jakub-pomykala replied:
Updated to use a StringBuilder

jhanders34 commented on dev/com.ibm.ws.kernel.reporting/src/com/ibm/ws/kernel/reporting/DataCollector.java:
```diff
@@ -0,0 +1,152 @@
+				// adding it.
+				if (!fd.getFeatureName().contains(":")) {
+					if (fd instanceof ProvisioningFeatureDefinition) {
+						if (!((ProvisioningFeatureDefinition) fd).isAutoFeature())
```
Unfortunately there is a public feature that is an auto feature.  It will enable even if the auto feature filter is not satisified.  As such I am not sure if this isAutoFeature conditional should be here.  Auto features are private or protected otherwise.

jakub-pomykala replied:
Could you confirm that we should remove is the `isAutoFeature()` check?

jhanders34 replied:
Yes I believe you should remove it.

jhanders34 commented on dev/com.ibm.ws.kernel.reporting/src/com/ibm/ws/kernel/reporting/DataCollectorException.java:
```diff
@@ -0,0 +1,27 @@
+	 * @throws NotSerializableException
+	 */
+	private void writeObject(ObjectOutputStream oos) throws NotSerializableException {
+		throw new NotSerializableException();
```
Not sure why we are focusing it to be no serializable.  There is nothing in the object that would prevent it from being serializable.

jakub-pomykala replied:
Removed

jhanders34 commented on dev/com.ibm.ws.kernel.reporting/src/com/ibm/ws/kernel/reporting/ReporterTask.java:
```diff
@@ -0,0 +1,102 @@
+	 * </pre>
+	 * 
+	 * @param e The Throwable
+	 * @return A string with the type and message of a Throwable and it's causes.
```
```suggestion
	 * @return A string with the type and message of a Throwable and its causes.
```

jhanders34 commented on dev/com.ibm.ws.kernel.reporting/src/com/ibm/ws/kernel/reporting/utils/CheckIfEnabledUtils.java:
```diff
@@ -0,0 +1,27 @@
+				enabled = false;
+			}
+		}
+		return enabled;
```
```suggestion
		return !Boolean.FALSE.equals(properties.get("enabled"));
```
No need to do containsKey if you switch the conditional to handle null.

jhanders34 commented on dev/com.ibm.ws.kernel.reporting/src/com/ibm/ws/kernel/reporting/utils/HashUtils.java:
```diff
@@ -0,0 +1,44 @@
+	 * @return A String
+	 * @throws DataCollectorException
+	 */
+	public static String hashString(String input) throws DataCollectorException {
```
Again a class with one method that is called from one class.  Not sure why you don't just have it in the class that calls it.

jakub-pomykala replied:
We are planning on calling this method in more than one place soon.

jhanders34 commented on dev/com.ibm.ws.kernel.reporting_fat/bnd.bnd:
```diff
@@ -0,0 +1,39 @@
+	com.ibm.ws.jmx.connector.client.restConnector;version=latest,\
+	io.openliberty.jakarta.restfulWS.3.1,\
+	com.ibm.websphere.javaee.jsonp.1.1,\
+	io.openliberty.jakarta.servlet.6.1,\
```
```suggestion
```

jhanders34 commented on dev/com.ibm.ws.kernel.reporting_fat/fat/src/com/ibm/ws/kernel/reporting/fat/CVEResponseTest.java:
```diff
@@ -0,0 +1,184 @@
+		if (server.isStarted()) {
+			server.stopServer();
+		}
+		if (testServer.isStarted()) {
```
might want this in a finally.  If the stop of the first server throws an exception, it will leave the second server running.

jakub-pomykala replied:
This is now in a finally

jhanders34 commented on dev/com.ibm.ws.kernel.reporting_fat/fat/src/com/ibm/ws/kernel/reporting/fat/CVEDataTest.java:
```diff
@@ -0,0 +1,130 @@
+package com.ibm.ws.kernel.reporting.fat;
```
Similarly the package name should start with io.openliberty.  You should also consider the project starting with io.oepnliberty as well, but I understand if you want to keep it by the other projects that start with com.ibm.ws.kernel projects.

jakub-pomykala replied:
Done.

helyarp commented on dev/com.ibm.ws.kernel.reporting/resources/com/ibm/ws/kernel/reporting/internal/resources/CVEReporting.nlsprops:
```diff
@@ -0,0 +1,59 @@
+CWWKF1700.reporting.is.enabled.useraction=To disabled CVE Reporting, specify <cveReporting enabled="false" /> in the server.xml file 
+
+CWWKF1701.reporting.is.disabled=CWWKF1701I: CVE Reporting feature is disabled
+CWWKF1701.reporting.is.disabled.explanation=CVE Reporting feature needs to be enabled to run
```
Change "needs to" to "must"

helyarp commented on dev/com.ibm.ws.kernel.reporting/resources/com/ibm/ws/kernel/reporting/internal/resources/CVEReporting.nlsprops:
```diff
@@ -0,0 +1,59 @@
+CWWKF1701.reporting.is.disabled.explanation=CVE Reporting feature needs to be enabled to run
+CWWKF1701.reporting.is.disabled.useraction=To enable CVE Reporting, specify <cveReporting enabled="true" /> in the server.xml file 
+
+open.liberty.resposne=For more information you can review Open Liberty CVEs at https://openliberty.io/docs/latest/security-vulnerabilities.html
```
Add comma after "For more information"

jakub-pomykala replied:
Done.

helyarp commented on dev/com.ibm.ws.kernel.reporting/resources/com/ibm/ws/kernel/reporting/internal/resources/CVEReporting.nlsprops:
```diff
@@ -0,0 +1,59 @@
+
+websphere.liberty.response=For more information you can review IBM WebSphere Liberty CVEs at https://www.ibm.com/support/pages/node/710969
+
+CWWKF1702.cve.found=CWWKF1702W: Based on a assessment of the Liberty runtime. It is recommended that following CVEs are reviewed: {0}
```
- Change "a assessment" to "an assessment" or "assessment"
- Change "runtime. " to "runtime, " (period to comma)
- Change "It is recommended that following CVEs are reviewed:" to "review the following CVEs:" or "it is recommended that you review the following CVEs:"

helyarp commented on dev/com.ibm.ws.kernel.reporting/resources/com/ibm/ws/kernel/reporting/internal/resources/CVEReporting.nlsprops:
```diff
@@ -0,0 +1,59 @@
+websphere.liberty.response=For more information you can review IBM WebSphere Liberty CVEs at https://www.ibm.com/support/pages/node/710969
+
+CWWKF1702.cve.found=CWWKF1702W: Based on a assessment of the Liberty runtime. It is recommended that following CVEs are reviewed: {0}
+CWWKF1702.cve.found.explanation=There is something that has been found
```
I assume that this message is "in progress." If this is the message, remove "There is" and change "that has been" to "was" ("Something was found.")

jakub-pomykala replied:
Updated. :)

helyarp commented on dev/com.ibm.ws.kernel.reporting/resources/com/ibm/ws/kernel/reporting/internal/resources/CVEReporting.nlsprops:
```diff
@@ -0,0 +1,59 @@
+CWWKF1702.cve.found.useraction=Follow the message to resolve what has been found
+
+CWWKF1703.cve.not.found=CWWKF1703I: Based on assessment of the Liberty runtime, no applicable CVEs were found.
+CWWKF1703.cve.not.found.explanation=There is no problems
```
Remove the "There is": "No problems were found."

jakub-pomykala replied:
Updated.

helyarp commented on dev/com.ibm.ws.kernel.reporting/resources/com/ibm/ws/kernel/reporting/internal/resources/CVEReporting.nlsprops:
```diff
@@ -0,0 +1,59 @@
+CWWKF1703.cve.not.found.explanation=There is no problems
+CWWKF1703.cve.not.found.useraction=The following message has no issues
+
+CWWKF1704.incorrect.url=CWWKF1704W: The URL specified is not correct and a connection could not be made: {0}
```
- (Optional) Change "URL specified" to "specified URL"
- (Required) Change "could not be" to "cannot be" or "was not"

jakub-pomykala replied:
Done.

helyarp commented on dev/com.ibm.ws.kernel.reporting/resources/com/ibm/ws/kernel/reporting/internal/resources/CVEReporting.nlsprops:
```diff
@@ -0,0 +1,59 @@
+CWWKF1703.cve.not.found.useraction=The following message has no issues
+
+CWWKF1704.incorrect.url=CWWKF1704W: The URL specified is not correct and a connection could not be made: {0}
+CWWKF1704.incorrect.url.explanation=Due to the URL provided no connection could be made
```
- (Required) Change "could be" to "was" or "can be" and add a comma after "provided"
- (Optional) To remove "Due to," change the message to "The provided URL did not connect" or "No connection can be made for the provided URL"
- (Global) Add period

jakub-pomykala replied:
Done.

helyarp commented on dev/com.ibm.ws.kernel.reporting/resources/com/ibm/ws/kernel/reporting/internal/resources/CVEReporting.nlsprops:
```diff
@@ -0,0 +1,59 @@
+
+CWWKF1706.issue.parsing=CWWKF1706W: Issue when parsing Product Information.
+CWWKF1706.issue.parsing.explanation=There was an issue when parsing the Product Information.
+CWWKF1706.issue.parsing.useraction=the service was not able to successfully parse the product information
```
- (Required) Change "the service" to "The service" and add a period at the end of the useraction statement.
- (Optional) Change "was not able to" to "cannot" or "did not"

jakub-pomykala replied:
Done.

helyarp reviewed (commented):
Suggested changes

jakub-pomykala committed:
Moving system props and beta checks into isEnabled()

- Moving these into one method
- Commenting out the beta check for now
- Adding AccessController.doPriviledged in order to get the system properties
- Renaming the system property

jakub-pomykala committed:
Updating project/package prefix to io.openliberty

Updating the messages

- Updating messages post review

- Updating message 1705

- removing whitespace in 1701

Updating the handling of response message

- Changing how the response message is built
- Adding in new test to check the response from the cloud service
- Refactoring variables containing "temp"
-

Updating the message file

- Removing un-used server from tests

Removing un-required useraction

Adding in useraction=None. for 1703

ProductEdition and ProductVersion changes

- Removing json4j from openidConnectClient as it is no longer required
- Changing test behaviour when it is stopping the servers
- Fixing the metatype.properties to match required and pass tests

Apply suggestions from code review

Co-authored-by: Jared Anderson <jhanders@us.ibm.com>

Replacing WebSphere with OpenLiberty in bundle

Co-authored-by: Jared Anderson <jhanders@us.ibm.com>

helyarp commented on dev/io.openliberty.reporting/resources/io/openliberty/reporting/internal/resources/CVEReporting.nlsprops:
```diff
@@ -0,0 +1,54 @@
+more.information.message=For more information, you can review {0}
+
+CWWKF1702.cve.found=CWWKF1702W: Based on an assessment of the Liberty runtime, it is recommended that you review the following CVEs: {0}
+CWWKF1702.cve.found.explanation=Security vulnerabilities have been found.
```
Change "have been" to "were"

helyarp commented on dev/io.openliberty.reporting/resources/io/openliberty/reporting/internal/resources/CVEReporting.nlsprops:
```diff
@@ -0,0 +1,54 @@
+
+CWWKF1702.cve.found=CWWKF1702W: Based on an assessment of the Liberty runtime, it is recommended that you review the following CVEs: {0}
+CWWKF1702.cve.found.explanation=Security vulnerabilities have been found.
+CWWKF1702.cve.found.useraction=Follow the message to resolve what has been found.
```
For the useraction, how about "Review the CVEs and resolve the security vulnerabilities." or ""Review the CVEs in the message and resolve the security vulnerabilities.""

helyarp commented on dev/io.openliberty.reporting/resources/io/openliberty/reporting/internal/resources/CVEReporting.nlsprops:
```diff
@@ -0,0 +1,54 @@
+CWWKF1700.reporting.is.enabled.useraction=To disable the checking for security vulnerabilities, specify <cveReporting enabled="false" /> in the server.xml file.
+
+CWWKF1701.reporting.is.disabled=CWWKF1701I: The checking for security vulnerabilities is disabled.
+CWWKF1701.reporting.is.disabled.explanation=The checking for security vulnerabilities is currently disabled.
```
For the explanation, use "The checking for security vulnerabilities must be enabled to run."

helyarp commented on dev/io.openliberty.reporting/resources/io/openliberty/reporting/internal/resources/CVEReporting.nlsprops:
```diff
@@ -0,0 +1,54 @@
+CWWKF1702.cve.found.useraction=Follow the message to resolve what has been found.
+
+CWWKF1703.cve.not.found=CWWKF1703I: Based on analysis of the Runtime, no security vulnerabilities requiring further analysis were found.
+CWWKF1703.cve.not.found.explanation=There is no security vulnerabilities found at this time.
```
Remove the "There is": "No security vulnerabilities were found at this time."

helyarp commented on dev/io.openliberty.reporting/resources/io/openliberty/reporting/internal/resources/CVEReporting.nlsprops:
```diff
@@ -0,0 +1,54 @@
+
+CWWKF1705.failed.response=CWWKF1705W: Unable to retrieve relevant CVE information due to a communication issue with the CVE Reporting Service.
+CWWKF1705.failed.response.explanation=The service was not able to give a response.
+CWWKF1705.failed.response.useraction=the service was not able to provide a response.
```
Change the 1705 explanation to "Fix the communication issue and try again."

helyarp commented on dev/io.openliberty.reporting/resources/io/openliberty/reporting/internal/resources/CVEReporting.nlsprops:
```diff
@@ -0,0 +1,54 @@
+
+CWWKF1706.issue.parsing=CWWKF1706W: Issue when parsing Product Information.
+CWWKF1706.issue.parsing.explanation=There was an issue when parsing the Product Information.
+CWWKF1706.issue.parsing.useraction=The service cannot successfully parse the product information.
```
The useraction "The service cannot ["did not"?] successfully parse the product information." would be better as the explanation.

Maybe use for the useraction "Fix the parsing issue and try again." or explain what the user needs to do.

helyarp reviewed (commented):
Suggested changes

jakub-pomykala commented:
#build

helyarp commented on dev/io.openliberty.reporting/resources/io/openliberty/reporting/internal/resources/CVEReporting.nlsprops:
```diff
@@ -0,0 +1,54 @@
+
+CWWKF1703.cve.not.found=CWWKF1703I: Based on analysis of the Runtime, no security vulnerabilities requiring further analysis were found.
+CWWKF1703.cve.not.found.explanation=No security vulnerabilities were found at this time.
+CWWKF1703.cve.not.found.useraction=The following message has no issues.
```
Remove "following"

jakub-pomykala replied:
Removed the whole useraction as it is not required.

jhanders34 replied:
I thought all messages needed a user action even if it was `None`.

jakub-pomykala replied:
Did a quick search and couldn't find a message without a useraction so I went ahead and added it back in with `None.`

helyarp reviewed (commented):
Suggested a change

helyarp reviewed (commented):
Looks good!

jakub-pomykala commented:
#build

jhanders34 commented on dev/io.openliberty.reporting/.project:
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<projectDescription>
+	<name>io.openliberty.reporting</name>
```
The project name should be `.internal` as part of the name.  `io.openliberty.reporting` would be the name of the bundle that would provide external APIs for applications to use.  See [Open Liberty naming guidelines](https://github.com/OpenLiberty/open-liberty/wiki/Open-Liberty-naming-guidelines) for additional details.

jakub-pomykala replied:
Done.

jhanders34 commented on dev/io.openliberty.reporting_fat/.project:
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<projectDescription>
+	<name>io.openliberty.reporting_fat</name>
```
Not explicitly required, but you may want to change it to `io.openliberty.reporting.internal_fat`

jakub-pomykala replied:
Done.

jhanders34 commented on dev/io.openliberty.reporting/src/io/openliberty/reporting/CVEResponseHandler.java:
```diff
@@ -0,0 +1,94 @@
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+
+package io.openliberty.reporting;
```
package name should include `.internal` since you are not exporting any of the classes in this bundle.

jakub-pomykala replied:
Done.

jhanders34 commented on dev/io.openliberty.reporting/bnd.bnd:
```diff
@@ -0,0 +1,52 @@
+
+-dsannotations: io.openliberty.reporting.FixReportingComponent
+
+instrument.classesExcludes: com/ibm/ws/kernel/reporting/internal/resources/*.class
```
```suggestion
instrument.classesExcludes: io/openliberty/reporting/internal/resources/*.class
```

jhanders34 commented on dev/io.openliberty.reporting/src/io/openliberty/reporting/CVEResponseHandler.java:
```diff
@@ -0,0 +1,94 @@
+				}
+				for (Entry<String, String> e : bulletin.entrySet()) {
+					StringBuilder msg = new StringBuilder().append(e.getKey()).append(" - ").append(e.getValue());
+					Tr.warning(tc, msg.toString());
```
This isn't taking a message key.  It shouldn't be done like this.  It seems there should be a `{1}` for CWWKF1702.cve.found and that this StringBuilder's output should be passed as a second parameter to the Tr.warning call above.  Or in reality it would move to here.

jakub-pomykala replied:
Done.

jhanders34 commented on dev/io.openliberty.reporting/src/io/openliberty/reporting/CVEServiceClient.java:
```diff
@@ -0,0 +1,173 @@
+	 * </p>
+	 * 
+	 * @param data    Map<String, String>
+	 * @param urlLink URL link which is set in the Server.xml
```
From the design call, the URL link is not supposed to be set in the server.xml.  Unless you are talking about proxy support in the future.

jakub-pomykala replied:
It is so that we can run tests without having to make a connection to the cloud service.

jhanders34 commented on dev/io.openliberty.reporting/src/io/openliberty/reporting/CVEServiceClient.java:
```diff
@@ -0,0 +1,173 @@
+		String jsonData = buildJsonString(data);
+
+		if (!urlLink.startsWith("https")) {
+			throw new MalformedURLException("Invalid Protocol, expected \"https\': " + urlLink);
```
```suggestion
			throw new MalformedURLException("Invalid protocol, expected https");
```

jhanders34 commented on dev/io.openliberty.reporting/src/io/openliberty/reporting/CVEServiceClient.java:
```diff
@@ -0,0 +1,173 @@
+	 */
+	public JSONObject retrieveCVEData(Map<String, String> data, String urlLink) throws IOException {
+
+		String jsonData = buildJsonString(data);
```
Your comment stated that this improves testability of the code to have this here instead of after the protocol check?  Can you explain?

jakub-pomykala replied:
If we move the `jsonData` to after the protocol check we won't be able to get this data within our tests. Because we have it here we are able to still run `sendData()` which allows us to check the data within our tests without a successful connection.

jhanders34 commented on dev/io.openliberty.reporting/src/io/openliberty/reporting/CVEServiceClient.java:
```diff
@@ -0,0 +1,173 @@
+	protected String buildJsonString(Map<String, String> data) {
+		StringBuilder jsonData = new StringBuilder("{");
+		for (String key : data.keySet()) {
+			if (key == "features" || key == "iFixes") {
```
You can't use `==` for Strings.  They are objects.  Either you need to use `if ("features".equals(key) || "iFixes".equals(key)) {` or you use a HashSet of Strings with "features" and "iFixes" in it and do a Set.contains(key) as the if conditional.

jakub-pomykala replied:
Done.

jhanders34 commented on dev/io.openliberty.reporting/src/io/openliberty/reporting/CVEServiceClient.java:
```diff
@@ -0,0 +1,173 @@
+						jsonData.append("\"").append(featureOrIfix).append("\", ");
+					}
+
+					jsonData.delete(jsonData.length() - 2, jsonData.length());
```
```suggestion
					jsonData.setLength(jsonData.length() - 2);
```

jhanders34 commented on dev/io.openliberty.reporting/src/io/openliberty/reporting/CVEServiceClient.java:
```diff
@@ -0,0 +1,173 @@
+				jsonData.append("\"").append(key).append("\": \"").append(data.get(key)).append("\", ");
+			}
+		}
+		jsonData.delete(jsonData.length() - 2, jsonData.length());
```
```suggestion
		jsonData.setLength(jsonData.length() - 2);
```

jhanders34 commented on dev/io.openliberty.reporting/src/io/openliberty/reporting/ReporterTask.java:
```diff
@@ -0,0 +1,112 @@
+
+	/**
+	 * <p>
+	 * Creates a string with the type and message of a Throwable and it's causes.
```
```suggestion
	 * Creates a string with the type and message of a Throwable and its causes.
```

jhanders34 commented on dev/io.openliberty.reporting/bnd.bnd:
```diff
@@ -0,0 +1,52 @@
+
+Bundle-Name: WebSphere kernel CVE Reporting
+Bundle-SymbolicName: io.openliberty.reporting
+Bundle-Description: WebSphere kernel CVE Reporting, version ${bVersion}
```
```suggestion
Bundle-Name: Liberty kernel CVE Reporting
Bundle-SymbolicName: io.openliberty.reporting
Bundle-Description: Liberty kernel CVE Reporting, version ${bVersion}
```

jhanders34 commented on dev/io.openliberty.reporting/resources/OSGI-INF/metatype/metatype.xml:
```diff
@@ -0,0 +1,41 @@
+
+ 
+    <!-- id gets mapped to configuration class impl in Designate element below --> 
+    <OCD id="io.openliberty.reporting.metatype" name="%cveReporting" description="%cveReporting.desc"
```
```suggestion
    <OCD id="io.openliberty.reporting.metatype" name="internal" description="%cveReporting.desc"
```

Until we are ready to beta / GA this it should be marked internal.

jhanders34 commented on dev/io.openliberty.reporting/src/io/openliberty/reporting/CVEServiceClient.java:
```diff
@@ -0,0 +1,173 @@
+			Tr.debug(this, tc, jsonData);
+		}
+
+		JSONObject json = JSONObject.parse(getResponse(connection));
```
```suggestion
		try (Reader reader = new InputStreamReader(connection.getInputStream(), StandardCharsets.UTF_8)) {
			JSONObject json = JSONObject.parse(reader);
		}
```

Can get rid of the getResponse method.

jhanders34 commented on dev/io.openliberty.reporting/src/io/openliberty/reporting/FixReportingComponent.java:
```diff
@@ -0,0 +1,141 @@
+		} else {
+			if (future != null && !future.isDone()) {
+				Tr.info(tc, "CWWKF1701.reporting.is.disabled");
+				future.cancel(false);
```
Should the future be nulled out after cancel as well?  Both here and in deactivate.

jakub-pomykala replied:
Done.

jakub-pomykala committed:
Rename bundle to internal

- Remove autoFeature check
- Null out future on deactive and modified (if disabled)
- Fix tests to check for any productEdition instead of just "open"
- Change condition for features or iFixes in CVEServiceClient
- Change the way the message us built

Fix for json4j causing test failures in WebSphere

- This condition will check if any of the bundles being added are "com.ibm..ws.json4j" before adding them to the content. if any of them are called "com.ibm..ws.json4j" it will be skipped.

Refactoring io.openliberty.reporting unittests pkg

Changing package name in kernelCore-1.0.mf file

- Adding in 2 more properties (os and osArch)
- Updating tests to match 2 new properties in the json being sent
- Updating the last of io.openliberty.reporting to io.openliberty.reporting.internal

Updating starting phase of com.ibm.json4j

Replacing tabs with spaces

Refactoring CVEResponseHandler class

- Removing internal from ConfigurationPid and factoryPid
- Adding in check for data.isEmpty to buildJsonString
- Removing conition for checking if feature is part of ProvisioningFeatureDefinition as no longer required since removing isAutoFeature check
- Updating 1702.cve.found message

ChangingCVEResponseHandler to be called statically

Co-authored-by: Jared Anderson <jhanders@us.ibm.com>

Removing extra colon

Co-authored-by: Jared Anderson <jhanders@us.ibm.com>

jakub-pomykala commented:
#build

jakub-pomykala commented:
#build

jakub-pomykala commented:
#build

jakub-pomykala commented:
#build

jhanders34 commented on dev/io.openliberty.reporting.internal/resources/OSGI-INF/metatype/metatype.xml:
```diff
@@ -0,0 +1,41 @@
+
+ 
+    <!-- id gets mapped to configuration class impl in Designate element below --> 
+    <OCD id="io.openliberty.reporting.internal.metatype" name="internal" description="%cveReporting.desc"
```
Not sure the id name and the factoryPid should have `.internal` in their name.  The metatype at least is an external name since this will eventually be a public metatype.  Not sure if factoryPid is exposed externally, but I would suggest removing the `.internal`.

jakub-pomykala replied:
Done.

jhanders34 commented on dev/io.openliberty.reporting.internal/src/io/openliberty/reporting/internal/CVEResponseHandler.java:
```diff
@@ -0,0 +1,112 @@
+
+	private static final TraceComponent tc = Tr.register(CVEResponseHandler.class);
+
+	private static final String LINE_SEPARATOR = getSystemProperty("line.separator");
```
```suggestion
	private static final String LINE_SEPARATOR = System.getProperty("line.separator");
```
`line.separator` is not a protected system property

jakub-pomykala replied:
Done.

jhanders34 commented on dev/io.openliberty.reporting.internal/src/io/openliberty/reporting/internal/CVEResponseHandler.java:
```diff
@@ -0,0 +1,112 @@
+			}
+		});
+	}
+
```
```suggestion
```

jakub-pomykala replied:
Done.

jhanders34 commented on dev/io.openliberty.reporting.internal/resources/io/openliberty/reporting/internal/resources/CVEReporting.nlsprops:
```diff
@@ -0,0 +1,54 @@
+
+more.information.message=For more information, you can review {0}
+
+CWWKF1702.cve.found=CWWKF1702W: Based on an assessment of {0}, review the following security bulletins:
```
```suggestion
CWWKF1702.cve.found=CWWKF1702W: Based on an assessment of {0}, review the following security bulletins: {1}
```

jakub-pomykala replied:
Done.

jhanders34 commented on dev/io.openliberty.reporting.internal/src/io/openliberty/reporting/internal/CVEResponseHandler.java:
```diff
@@ -0,0 +1,112 @@
+
+				JSONArray cvesArray = new JSONArray();
+
+				StringBuilder message = new StringBuilder();
+
+				message.append(Tr.formatMessage(tc, "CWWKF1702.cve.found", cveProduct.get("productName")));
+
```
```suggestion
```

jakub-pomykala replied:
Done.

jhanders34 commented on dev/io.openliberty.reporting.internal/src/io/openliberty/reporting/internal/CVEResponseHandler.java:
```diff
@@ -0,0 +1,112 @@
+
+				}
+
+				for (Entry<String, String> e : bulletin.entrySet()) {
+					message.append(LINE_SEPARATOR);
+					message.append(e.getKey()).append(" - ").append(e.getValue());
+				}
+
+				Tr.warning(tc, message.toString());
```
```suggestion
				StringBuilder message = new StringBuilder();
				for (Entry<String, String> e : bulletin.entrySet()) {
					message.append(LINE_SEPARATOR);
					message.append(e.getKey()).append(" - ").append(e.getValue());
				}

				Tr.warning(tc, "CWWKF1702.cve.found", cveProduct.get("productName"), message.toString());
```

jakub-pomykala replied:
Done .

jhanders34 commented on dev/io.openliberty.reporting.internal/src/io/openliberty/reporting/internal/CVEServiceClient.java:
```diff
@@ -0,0 +1,155 @@
+				jsonData.append("\"").append(key).append("\": \"").append(data.get(key)).append("\", ");
+			}
+		}
+		jsonData.setLength(jsonData.length() - 2);
```
The code assumes that data is not an empty Map.  It appears that that assumption is valid from the callers of the code, but if you want to make the code more robust you may want to add a `if (!data.isEmpty())` around the setLength method or have the first thing in the method be:

```
		if (data.isEmpty()) {
			return "{}";
		}
```

jakub-pomykala replied:
Added `if (data.isEmpty()` to the top of the method.

jhanders34 commented on dev/io.openliberty.reporting.internal/src/io/openliberty/reporting/internal/DataCollector.java:
```diff
@@ -0,0 +1,165 @@
+				productVersion = allProductInfo.get("com.ibm.websphere.appserver").getVersion();
+				productEdition = allProductInfo.get("com.ibm.websphere.appserver").getEdition();
+			} else {
+				productVersion = allProductInfo.get("io.openliberty").getVersion();
+				productEdition = allProductInfo.get("io.openliberty").getEdition();
```
This assumes that allProductInfo will have `"io.openliberty"` in it.  If it doesn't this will be a NullPointerException.  Is it guaranteed to have io.openliberty in it?

jakub-pomykala replied:
As far as I am aware if the product is not WebSphere Liberty it will have to be Open Liberty.

jhanders34 commented on dev/io.openliberty.reporting.internal/src/io/openliberty/reporting/internal/DataCollector.java:
```diff
@@ -0,0 +1,165 @@
+				// If we don't want to include auto features, then check each feature before
+				// adding it.
+				if (!fd.getFeatureName().contains(":")) {
+					if (fd instanceof ProvisioningFeatureDefinition) {
+						publicFeatures.add(fd.getFeatureName());
+
+					} else {
+						// If we're not an instance of ProvisioningFeatureDefinition then add the
+						// feature to the list.
+						publicFeatures.add(fd.getFeatureName());
+					}
```
```suggestion
					publicFeatures.add(fd.getFeatureName());
```

This gets much simpler now.

jakub-pomykala replied:
Done.

jhanders34 commented on dev/io.openliberty.reporting.internal/src/io/openliberty/reporting/internal/FixReportingComponent.java:
```diff
@@ -0,0 +1,143 @@
+ * The main entry point into CVE Reporting service
+ * </p>
+ */
+@Component(immediate = true, configurationPid = "io.openliberty.reporting.internal.CVEReportingComponent", configurationPolicy = ConfigurationPolicy.OPTIONAL)
```
If you remove the `.internal` from the metatype, reminder to move it here as well for the `configurationPid`

jakub-pomykala replied:
Done.

jhanders34 commented on dev/io.openliberty.reporting.internal/src/io/openliberty/reporting/internal/ReporterTask.java:
```diff
@@ -0,0 +1,112 @@
+			Map<String, String> data = collector.getData();
+			String urlLink = (String) props.get("urlLink");
+			JSONObject response = new CVEServiceClient().retrieveCVEData(data, urlLink);
+			new CVEResponseHandler().handleResponse(data.get("productEdition"), response);
```
CVEResponseHandler is another class now with one method in it.  Could that method be moved to this class.  If you choose not to do that, I would make that method `static` and not have to new up a CVEResponseHandler for no reason.

jakub-pomykala replied:
Prefer to keep it in its own class but so I made the method `static`.

jhanders34 reviewed (changes requested):
In general, all of this code is written with tab indentation instead of spaces.  This really makes it difficult with line wrapping when looking at the code because of 8 character tabs.  I would recommend updating your IDE to use 4 spaces instead of tabs and to update all of the code in the reporting and reporting fat project to the new format.

jakub-pomykala commented:
#build

jhanders34 commented on dev/io.openliberty.reporting.internal/src/io/openliberty/reporting/internal/ReporterTask.java:
```diff
@@ -0,0 +1,112 @@
+            Map<String, String> data = collector.getData();
+            String urlLink = (String) props.get("urlLink");
+            JSONObject response = new CVEServiceClient().retrieveCVEData(data, urlLink);
+            new CVEResponseHandler().handleResponse(data.get("productEdition"), response);
```
```suggestion
            CVEResponseHandler.handleResponse(data.get("productEdition"), response);
```
The point of making it static was so you would call it statically.

jakub-pomykala replied:
I thought I made this change!

jhanders34 commented on dev/io.openliberty.reporting.internal/resources/io/openliberty/reporting/internal/resources/CVEReporting.nlsprops:
```diff
@@ -0,0 +1,54 @@
+
+more.information.message=For more information, you can review {0}
+
+CWWKF1702.cve.found=CWWKF1702W: Based on an assessment of {0}, review the following security bulletins: : {1}
```
```suggestion
CWWKF1702.cve.found=CWWKF1702W: Based on an assessment of {0}, review the following security bulletins: {1}
```
Extra colon not needed.

jakub-pomykala commented:
#libby

jakub-pomykala commented:
error/failures in build are not related to code in this PR.

wraschke commented:
This causes a failure in packaging verification:

```
Message verification failed! A translated message file is missing from one or more bundle jars. If this failure is due to one or more new message files that have not yet been translated, or to the inclusion of a thirdpary bundle jar, please update the list of ignored failures in ant_build/resources/configuration/messageVerifyIgnores. The gradle failure is: 
 
FAILURE: Build failed with an exception.

* Where:
Build file '/home/jazz_build/Build/jbe/build/dev/cnf/gradle/scripts/verification.gradle' line: 62

* What went wrong:
Execution failed for task ':verifyBundleNLS'.
> Missing metatype.properties files: io.openliberty.reporting.internal_1.0.89.jar

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org.

BUILD FAILED in 39s
```
